Upgrade to OpenMOPAC

JacksonBurns opened the pull request:
The current version of MOPAC used by RMG is semi-closed source and requires an key. The new version, OpenMOPAC, is freely available and maintained by the community.

This PR will attempt to switch to this newer version. Doing so will fix some issues on the user side (see below) and also allow forked repositories to run the continuous integration testing (which currently fails due to the requirement for a license key, not present on forks).

Resolves #2328 which originally discussed the possibility of moving to OpenMOPAC.
Resolves #2251 which was facing issues with running the outdated version of MOPAC and attempted to install this new version without success.
Resolves #2407 also facing licensing issues.

PR #2322 attempted to do this in the past but also attempted to upgrade other packages and ran into difficulties; this PR will only address MOPAC->OpenMOPAC and those packages that are _directly_ affected. This PR seemed to also be blocked by #2272 which fixed an issue with RDKit (see [this comment](https://github.com/ReactionMechanismGenerator/RMG-Py/issues/2328#issuecomment-1335628473) from @xiaoruiDong) that has since been merged.

PR #2386 also attempted to loosen the environment restrictions but was blocked by other PRs that added support for updated versions of Cantera.

hwpang reviewed (approved):
Looks good to me. Would like @rwest's comments in this before merging.

rwest reviewed (commented):
Haven't tried compiling or running (but I presume you have). Just read it on github. Looks ok. Please can you push the binaries to the official rmg channel and refer to that in the conda environment file?

JacksonBurns commented on environment.yml:
```diff
@@ -34,9 +35,9 @@ dependencies:
   - pandas
   - psutil
-  - rmg::pydas >=1.0.2
+  - jacksonburns::pydas >=1.0.3
```
I will move them over now, as well as build on a mac.

The recipes are in the ReactionMechanismGenerator/PyD* repos, but are not being built automatically

JacksonBurns commented on rmgpy/qm/mopac.py:
```diff
@@ -45,100 +45,72 @@
+        mopacEnv = os.getenv("MOPAC_DIR", default="/opt/mopac")
+        executablePath = os.path.join(mopacEnv, "mopac")
+        if not os.path.exists(executablePath):
+            executablePath = os.path.join(mopacEnv, "(MOPAC 2009 or 2012 or 2016)")
```
This entire block does not make sense to me, I left it in to spark some discussion.

I think in the previous versions we may have allowed users to specify a different MOPAC executable and then retrieved it from there, but this isn't something we are interested in now since we ship with the latest release.

I will just remove this. It's dead and untested anyway.

JacksonBurns commented:
@rwest I made the requested changes, I uploaded PyDAS and PyDQED for Linux to the RMG channel, and @hwpang and I uploaded the conda packages for PyDAS and PyDQED for MacOS.

Most of the edits shown in the diff are just from applying PEP8.

rwest commented:
Regression test error `TypeError: quantum_mechanics() got an unexpected keyword argument 'fileStore'` ?

rwest commented:
I think I see the problem. Your `gigantic PEP8 refactor for all things touching QMSettings` commit has changed some things that are part of input file syntax, not just internal variable names. I think we shouldn't make breaking changes like this to an input file format just for style consistency, especially through blindly applying an automated tool, on a PR that's unrelated.

To preserve things in case I messed up I have pushed your branch to a new one called `new-mopac-backup` which we can delete once everything is resolved.   I have then rebased a few times, merging some commits, trying to tidy up a bit, and dropped the gigantic PEP8 refactor commit.   I have force-pushed the result to this `new-mopac` branch, and we'll see if the test suite runs.  If I've messed up, feel free to revert to `new-mopac-backup` and try to fix things yourself.

JacksonBurns commented:
@rwest thanks for resolving that - I tried to get all the instances fixed but apparently missed some. Agreed that it would be a bad idea to change input file format for something like this. Thanks for taking care of the fix! Should pass.

rwest commented:
Not yet sure if it's the changes on this branch, or a problem already in `main`, but making a new conda environment on my mac is faling:

```
git:(new-mopac) ✗ conda env create -f environment.yml 
Channels:
 - defaults
 - rmg
 - conda-forge
 - cantera
 - pr-omethe-us
 - kyleniemeyer
Platform: osx-64
Collecting package metadata (repodata.json): done
Solving environment: | error    libmamba Selected channel specific (or force-reinstall) job, but package is not available from channel. Solve job will fail.
- error    libmamba Selected channel specific (or force-reinstall) job, but package is not available from channel. Solve job will fail.
failed
Channels:
 - defaults
 - rmg
 - conda-forge
 - cantera
 - pr-omethe-us
 - kyleniemeyer
Platform: osx-64
Collecting package metadata (repodata.json): done
Solving environment: \ error    libmamba Selected channel specific (or force-reinstall) job, but package is not available from channel. Solve job will fail.
| error    libmamba Selected channel specific (or force-reinstall) job, but package is not available from channel. Solve job will fail.
failed

LibMambaUnsatisfiableError: Encountered problems while solving:
  - unsupported request
```

hwpang commented:
@rwest 

This is probably because of this package version specification 

```
conda-forge::libgfortran-ng>=10.0
```

JacksonBurns commented:
> Not yet sure if it's the changes on this branch, or a problem already in `main`, but making a new conda environment on my mac is faling:
> 
> ```
> git:(new-mopac) ✗ conda env create -f environment.yml 
> Channels:
>  - defaults
>  - rmg
>  - conda-forge
>  - cantera
>  - pr-omethe-us
>  - kyleniemeyer
> Platform: osx-64
> Collecting package metadata (repodata.json): done
> Solving environment: | error    libmamba Selected channel specific (or force-reinstall) job, but package is not available from channel. Solve job will fail.
> - error    libmamba Selected channel specific (or force-reinstall) job, but package is not available from channel. Solve job will fail.
> failed
> Channels:
>  - defaults
>  - rmg
>  - conda-forge
>  - cantera
>  - pr-omethe-us
>  - kyleniemeyer
> Platform: osx-64
> Collecting package metadata (repodata.json): done
> Solving environment: \ error    libmamba Selected channel specific (or force-reinstall) job, but package is not available from channel. Solve job will fail.
> | error    libmamba Selected channel specific (or force-reinstall) job, but package is not available from channel. Solve job will fail.
> failed
> 
> LibMambaUnsatisfiableError: Encountered problems while solving:
>   - unsupported request
> ```

Can you try deleting `  - conda-forge::libgfortran-ng>=10.0` from the environment file and running again? @hwpang and I ran into a problem while building PyDAS and PyDQED on MacOS yesterday that this package is not available on mac, but that conda finds a different one that fits the requirements. I had only added this to the list to be explicit, but I only ever dev on ubuntu.

rwest commented:
I removed that line as you suggest. It seems to solve OK on my mac (though haven't yet checked the resulting environment works). Have pushed it to see if the CI tests pass.

rwest commented:
This looks promising. Seems to build on my mac, and the CI tests haven't failed.
But we updated these in the environment.yml
```
   - rmg::pydas >=1.0.3
   - rmg::pydqed >=1.0.3
 ```
but the `.conda/meta.yaml` still has
```
    - pydas >=1.0.2
    - pydqed >=1.0.1
 ```
should that be updated too? Any other changes to conda stuff?
(I'm not up to date with how the conda binary building works)

JacksonBurns commented:
> This looks promising. Seems to build on my mac, and the CI tests haven't failed.
> But we updated these in the environment.yml
> ```
>    - rmg::pydas >=1.0.3
>    - rmg::pydqed >=1.0.3
>  ```
> but the `.conda/meta.yaml` still has
> ```
>     - pydas >=1.0.2
>     - pydqed >=1.0.1
>  ```
> should that be updated too? Any other changes to conda stuff?
> (I'm not up to date with how the conda binary building works)

I think they *should* be updated, I also don't know anything about actually building the RMG binaries - but we don't need to build anaconda RMG binaries anymore since we moved to docker. We could update them for now, with a separate PR to remove the file (likely .conda as a whole)?

rwest commented:
> >  Any other changes to conda stuff?
> 
> I think they _should_ be updated, I also don't know anything about actually building the RMG binaries - but we don't need to build anaconda RMG binaries anymore since we moved to docker. We could update them for now, with a separate PR to remove the file (likely .conda as a whole)?

Let's update the `.conda` folder in a separate PR. I notice the cantera version got left behind too. Then this PR is more nearly ready to merge (as soon as tests pass?)

JacksonBurns commented:
> > >  Any other changes to conda stuff?
> > 
> > I think they _should_ be updated, I also don't know anything about actually building the RMG binaries - but we don't need to build anaconda RMG binaries anymore since we moved to docker. We could update them for now, with a separate PR to remove the file (likely .conda as a whole)?
> 
> Let's update the `.conda` folder in a separate PR. I notice the cantera version got left behind too. Then this PR is more nearly ready to merge (as soon as tests pass?)

Ok, I will open a PR to update .conda following this. You are correct, we missed this in the Canter 2.6 PR as well.

Yes, assuming this passes it can be merged!

rwest commented:
A slight confusion - at least confused me for a while - is that the new pydqed 1.0.3 binary you built reports that it is 1.0.1 when you run `make check`. 

```
(rmg_env) ➜  RMG-Py git:(new-mopac) ✗ conda list pydqed
# packages in environment at /opt/miniconda3/envs/rmg_env:
#
# Name                    Version                   Build  Channel
pydqed                    1.0.3            py37h41275ba_6    rmg

(rmg_env) ➜  RMG-Py git:(new-mopac) ✗ make check

Checking vital dependencies...

Package        Version        Location
lpsolve55                     /opt/miniconda3/envs/rmg_env/lib/python3.7/site-packages/lpsolve55.cpython-37m-darwin.so
OpenBabel      3.1.0          /opt/miniconda3/envs/rmg_env/lib/python3.7/site-packages/openbabel/openbabel.py
pydqed         1.0.1          /opt/miniconda3/envs/rmg_env/lib/python3.7/site-packages/pydqed.cpython-37m-darwin.so
pyrdl                         /opt/miniconda3/envs/rmg_env/lib/python3.7/site-packages/py_rdl/__init__.py
RDKit          2020.03.3      /opt/miniconda3/envs/rmg_env/lib/python3.7/site-packages/rdkit/__init__.py
symmetry       1.16           /opt/miniconda3/envs/rmg_env/bin/symmetry

Everything was found :)


```

hwpang commented:
> A slight confusion - at least confused me for a while - is that the new pydqed 1.0.3 binary you built reports that it is 1.0.1 when you run `make check`.
> 
> ```
> (rmg_env) ➜  RMG-Py git:(new-mopac) ✗ conda list pydqed
> # packages in environment at /opt/miniconda3/envs/rmg_env:
> #
> # Name                    Version                   Build  Channel
> pydqed                    1.0.3            py37h41275ba_6    rmg
> 
> (rmg_env) ➜  RMG-Py git:(new-mopac) ✗ make check
> 
> Checking vital dependencies...
> 
> Package        Version        Location
> lpsolve55                     /opt/miniconda3/envs/rmg_env/lib/python3.7/site-packages/lpsolve55.cpython-37m-darwin.so
> OpenBabel      3.1.0          /opt/miniconda3/envs/rmg_env/lib/python3.7/site-packages/openbabel/openbabel.py
> pydqed         1.0.1          /opt/miniconda3/envs/rmg_env/lib/python3.7/site-packages/pydqed.cpython-37m-darwin.so
> pyrdl                         /opt/miniconda3/envs/rmg_env/lib/python3.7/site-packages/py_rdl/__init__.py
> RDKit          2020.03.3      /opt/miniconda3/envs/rmg_env/lib/python3.7/site-packages/rdkit/__init__.py
> symmetry       1.16           /opt/miniconda3/envs/rmg_env/bin/symmetry
> 
> Everything was found :)
> ```

We probably need to update this line: https://github.com/ReactionMechanismGenerator/PyDQED/blob/libfortran-update/pydqed.pyx#L48 to 1.0.3

JacksonBurns commented:
The anaconda package is 1.0.3 but we made no changes to PyDQED itself so it still shows the old version. We can update the version in PyDQED to match the conda package, but it does not affect performance.

rwest commented on rmgpy/qm/gaussian.py:
```diff
@@ -44,68 +44,78 @@
+    output_file_extension = ".log"
 
-    executablesToTry = ('g16', 'g09', 'g03')
+    executablesToTry = ("g16", "g09", "g03")
```
if we're going through all the code churn of running black on the file and changing ' to " etc., should we also change `executablesToTry` and `possibleDirs` (below) to snake_case?
(I'm not a huge far of the un-necessarry code churn of black).

rwest reviewed (approved):
Not a huge fan of the code churn from autoformatting, but there's no good answer really. At least once it's done it's done.  I noticed one inconsistency but it's minor. I think this is fine to merge, in my opinion.
